File: src/main/java/com/github/hhhzzzsss/songplayer/playing/SongQueue.java
package com.github.hhhzzzsss.songplayer.playing;

import com.github.hhhzzzsss.songplayer.SongPlayer;
import com.github.hhhzzzsss.songplayer.song.Song;
import com.github.hhhzzzsss.songplayer.song.SongLoaderThread;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

public class SongQueue {
    private final LinkedList<Song> songQueue = new LinkedList<>();
    private Song currentSong = null;

    private SongLoaderThread loaderThread = null;

    // Playback
    public void next() {
        currentSong = null;
        checkSong();
    }

    public void clear() {
        currentSong = null;
        clearQueue();
    }

    public void clearQueue() {
        songQueue.clear();
    }

    // Accessors
    public List<Song> getQueue() {
        return List.copyOf(songQueue);
    }

    public Song getCurrentSong() {
        return currentSong;
    }

    // Loading
    public void addSong(Song song) {
        songQueue.add(song);
        checkSong();
    }

    public void loadSong(String location) throws IOException, IllegalStateException {
        if (loaderThread != null) {
            throw new IllegalStateException("Already loading a song, cannot load another");
        }

        loaderThread = new SongLoaderThread(location);
        loaderThread.start();
    }

    public void loadSong(SongLoaderThread thread) throws IllegalStateException {
        if (loaderThread != null) {
            throw new IllegalStateException("Already loading a song, cannot load another");
        }

        loaderThread = thread;
    }

    public void checkLoaderThread() {
        if (loaderThread == null || loaderThread.isAlive()) return;

        if (loaderThread.exception != null) {
            SongPlayer.addChatMessage("§cFailed to load song: §4" + loaderThread.exception.getMessage());
        } else {
            addSong(loaderThread.song);
        }

        loaderThread = null;
    }

    private void checkSong() {
        if (currentSong == null) currentSong = songQueue.pollFirst();
    }

    // Other
    public boolean isEmpty() {
        return currentSong == null && songQueue.isEmpty();
    }
}

File: src/main/java/com/github/hhhzzzsss/songplayer/playing/SongQueue.kt
package com.github.hhhzzzsss.songplayer.playing

import com.github.hhhzzzsss.songplayer.SongPlayer
import com.github.hhhzzzsss.songplayer.song.Song
import com.github.hhhzzzsss.songplayer.song.SongLoaderThread
import java.io.IOException
import java.util.*

class SongQueue {
    private val songQueue = mutableListOf<Song>()

    var currentSong: Song? = null
        private set

    private var loaderThread: SongLoaderThread? = null

    // Playback
    fun forcePlay(song: Song) {
        currentSong = song
    }

    fun next() {
        currentSong = null
        checkSong()
    }

    fun clear() {
        currentSong = null
        songQueue.clear()
    }

    val queue: Array<Song>
        // Accessors
        get() = songQueue.toTypedArray()

    // Loading
    fun addSong(song: Song) {
        songQueue.add(song)
        checkSong()
    }

    @Throws(IOException::class, IllegalStateException::class)
    fun loadSong(location: String) {
        check(loaderThread == null) { "Already loading a song, cannot load another" }

        loaderThread = SongLoaderThread(location)
        loaderThread!!.start()
    }

    @Throws(IllegalStateException::class)
    fun loadSong(thread: SongLoaderThread) {
        check(loaderThread == null) { "Already loading a song, cannot load another" }

        loaderThread = thread
    }

    fun checkLoaderThread() {
        val lt = loaderThread ?: return
        if (lt.isAlive) return

        if (lt.exception != null) {
            SongPlayer.addChatMessage("§cFailed to load song: §4" + lt.exception.message)
            loaderThread = null
            return
        }

        addSong(lt.song)
        loaderThread = null
    }

    private fun checkSong() {
        if (queue.isEmpty()) return
        if (currentSong == null) currentSong = songQueue.removeFirst()
    }

    val isEmpty: Boolean
        get() = currentSong == null && songQueue.isEmpty()
}